docs/ProofOptimization/ProofOptimization.md

```diff
@@ -0,0 +1,599 @@
+
+## Avoiding quantifiers
+
+Dafny allows you to prove both quantified and unquantified version of a fact. And, in some cases, the quantified version can be more convenient, allowing you to prove things with fewer steps, and with less need to keep track of details. For example, the Dafny `libraries` repository contains a definition that looks roughly like the following (with a few simplifications).
+
+<!-- %no-check -->
+```dafny
+lemma LemmaMulAuto()
+  ensures forall x:int, y:int :: x * y == y * x
+  ensures forall x:int, y:int, z:int :: (x + y) * z == x * z + y * z
+  ensures forall x:int, y:int, z:int :: (x - y) * z == x * z - y * z
+{
+...
+}
+  
+lemma LemmaMulIsDistributiveAddOtherWay(x: int, y: int, z: int)
+  ensures (y + z) * x == y * x + z * x
+{
+  LemmaMulAuto();
+}
+```
+
+However, although these proofs are very succinct, they require the verifier to do more work. If you encounter cases where this leads to verification failures, it can often help to prove more specific instances. 
+
+One particular pattern that works well when creating general-purpose lemmas proving universal properties is to prove a parameterized version first and then use that to prove the universal version. Then, clients of the lemma can choose one or the other.
+
+For example, consider the following two lemmas taken from the `libraries` repository.
+
+<!-- %check-verify -->
+```dafny
+lemma LemmaMulEquality(x: int, y: int, z: int)
+  requires x == y
+  ensures x * z == y * z
+{}
+
+lemma LemmaMulEqualityAuto()
+  ensures forall x: int, y: int, z: int :: x == y ==> x * z == y * z
+{
+  forall (x: int, y: int, z: int | x == y)
+    ensures x * z == y * z
+  {
+    LemmaMulEquality(x, y, z);
+  }
+}
+```
+
+There was another instance of code in the repository where a proof began to fail after upgrading the version of Z3 included with Dafny. That code originally used `LemmaMulEqualityAuto`. We [updated it](https://github.com/dafny-lang/libraries/blob/7c386fa0b4a267715f9bd49948d1af68e1631e6b/src/dafny/Collections/LittleEndianNat.dfy#L532) to call `LemmaMulEquality` with the appropriate parameters and the proof reliably succeeded.
+
+In general, creating and using explicit lemmas like `LemmaMulEquality`, and skipping lemmas like `LemmaMulEqualityAuto` altogether, tends to lead to proofs with less variability. 
+
+## Adding triggers
```

Comment:
I would explicitly flag this as a last resort, and perhaps even say if you find you HAVE to use triggers you should reach out to us to provide a better solution.